smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TracingSymbolProvider.java

```diff
@@ -0,0 +1,192 @@
+         *                            TraceFile.
+         * @return This Builder.
+         */
+        public Builder artifactDefinitions(ArtifactDefinitions artifactDefinitions) {
+            this.artifactDefinitions = artifactDefinitions;
+            return this;
+        }
+
+        /**
+         * Sets this Builder's ArtifactMetadata.
+         *
+         * @param artifactMetadata ArtifactMetadata for this TracingSymbolProvider's
+         *                         TraceFile.
+         * @return This Builder.
+         */
+        public Builder artifactMetadata(ArtifactMetadata artifactMetadata) {
+            this.artifactMetadata = artifactMetadata;
+            return this;
+        }
+
+        /**
+         * Sets this Builder's shapeLinkCreator. The shapeLinkCreator
+         * is a function that maps from a Symbol to a List of ShapeLinks.
+         * Custom Functions should be designed for each code generator
+         * that map apply the tags and types in the definitions files
+         * to specific ShapeLinks.
+         *
+         * @param shapeLinkCreator A Function that defines a mapping
+         *                         from a Symbol to a List of ShapeLinks.
+         * @return This Builder.
+         */
+        public Builder shapeLinkCreator(BiFunction<Shape, Symbol, List<ShapeLink>> shapeLinkCreator) {
+            this.shapeLinkCreator = shapeLinkCreator;
+            return this;
+        }
+
+        /**
+         * Sets this Builder's SymbolProvider.
+         *
+         * @param symbolProvider The SymbolProvider that the
+         *                       TracingSymbolProvider will decorate.
+         * @return This Builder.
+         */
+        public Builder symbolProvider(SymbolProvider symbolProvider) {
+            this.symbolProvider = symbolProvider;
+            return this;
+        }
+
+        /**
+         * Sets the Builder's {@link ArtifactMetadata} based on the given type and
+         * default values for other required fields. This method should ONLY be used
+         * when the version, type, homepage, and typeVersion of the ArtifactMetadata
+         * object is unknown at the time of code generation. This method will ONLY set
+         * the required fields of the {@link ArtifactMetadata}.
+         *
+         * <p>The type is set to the artifactType that is passed in. The artifactType is
+         * the code language of the generated artifact, e.g. Java.
+         *
+         * <p>The timestamp in ArtifactMetadata is set to the current time when the
+         * method is called.
+         *
+         * <p>The id and version are set to a UUID that should be changed after the
+         * TraceFile is constructed and the correct id and version are known.
+         *
+         * @param artifactType The type, i.e. language, of the ArtifactMetadata object.
+         * @return This Builder.
+         */
+        public Builder setArtifactMetadataAsDefault(String artifactType) {
```

Comment:
Can this method be located alongside the other method that sets `artifactMetadata`?